### PR TITLE
Pipelines: Add first reserved Step operation "Build"

### PR DESCRIPTION
**Pull request by briancain**
This pull request introduces the first implementation of reserved Step operations "Build". It accepts a single option for whether or not the build operation should push its artifact. This matches the BuildOp job proto as well as the configurable option when running `waypoint build`.

### CODE REVIEW

File: internal/config/pipeline.go [hunk 1]
```diff
@@ -127,11 +127,14 @@ func (c *Config) PipelineProtos() ([]*pb.Pipeline, error) {
 			},
 		}
 
+		// TODO(briancain): Use the default ODR image for image_url so that it isn't required
+		// We do this already maybe in the ODR task launcher.
```
Comment (mitchellh) [inline internal/config/pipeline.go:131]: Yeah, you can just make it empty and I think we should pick up the default that is configured with the server... that maybe won't work today but that's probably how it should work.

File: internal/runner/operation_pipeline_step.go [hunk 2]
```diff
@@ -29,67 +29,26 @@ func (r *Runner) executePipelineStepOp(
 	case *pb.Pipeline_Step_Exec_:
 		return r.executePipelineStepExec(ctx, log, job, project, kind.Exec)
 
+	case *pb.Pipeline_Step_Build_:
```
Comment (mitchellh) [inline internal/runner/operation_pipeline_step.go:32]: Ah! I would actually never hit this. I would change `RunPipeline` on the server to transform this into a build job _there_ so we never have to spin up this superflous ODR/task.

**Review by mitchellh — CHANGES_REQUESTED** [review]

File: internal/runner/operation_pipeline_step.go (hunk 2 above)
Comment (briancain) [reply internal/runner/operation_pipeline_step.go:32]: Ahhh ok, I'll look into making that work 👍🏻

Comment (briancain) [reply internal/runner/operation_pipeline_step.go]: Got that working in [b4a036a](https://github.com/hashicorp/waypoint/pull/3459/commits/b4a036a2350837512f4d951f6ce907735f1ef319)

**Review by mitchellh — APPROVED** [review]
This looks great!

**Pull request merged**